README.md
# ubuntu-18.04-chartmuseum

Vagrant Box with Ubuntu 18.04 and chartmuseum

## Base image

Used base image [elegoev/ubuntu-18.04-docker](https://app.vagrantup.com/elegoev/boxes/ubuntu-18.04-docker)

## Directory Description

| directory | description                                          |
|-----------|------------------------------------------------------|
| inspec    | inspec test profiles with controls                   |
| packer    | packer build, provisioner and post-processor scripts |
| test      | test environment for provision & inspec development  |

## Configuration

### Vagrant Cloud

- [elegoev/ubuntu-18.04-chartmuseum](https://app.vagrantup.com/elegoev/boxes/ubuntu-18.04-chartmuseum)

### Useful Vagrant Plugins

- [vagrant-disksize](https://github.com/sprotheroe/vagrant-disksize)
- [vagrant-hosts](https://github.com/oscar-stack/vagrant-hosts)
- [vagrant-secret](https://github.com/tcnksm/vagrant-secret)
- [vagrant-vbguest](https://github.com/dotless-de/vagrant-vbguest)
- [vagrant-serverspec](https://github.com/vvchik/vagrant-serverspec)
- [vagrant-vmware-esxi](https://github.com/josenk/vagrant-vmware-esxi)

### Using chartmuseum

#### Vagrantfile

    Vagrant.configure("2") do |config|

        $basebox_name="ubuntu-18.04-chartmuseum-test"
        $basebox_hostname="ubuntu-1804-chartmuseum-test"
        $src_image_name="elegoev/ubuntu-18.04-chartmuseum"
        $vb_group_name="basebox-chartmuseum-test"

        config.vm.define "#{$basebox_name}" do |machine|
          machine.vm.box = "#{$src_image_name}"
    
          # define guest hostname
          machine.vm.hostname = "#{$basebox_hostname}"

          machine.vm.provider "virtualbox" do |vb|
            vb.name = $basebox_name
            vb.cpus = 1
            vb.customize ["modifyvm", :id, "--memory", "1024" ]
            vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
            vb.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
            vb.customize ["modifyvm", :id, "--groups", "/#{$vb_group_name}" ]
            vb.customize ["modifyvm", :id, "--vram", 256 ]
          end

        end   

    end

### Refrenzen

- [Chartmuseum](https://chartmuseum.com/)
- [Docker Hub Chartmuseum](https://hub.docker.com/u/chartmuseum)
- [Upload Charts](https://chartmuseum.com/docs/#uploading-a-chart-package)

### Versioning

Repository follows sematic versioning  [![semantic versioning](https://img.shields.io/badge/semver-2.0.0-green.svg)](http://semver.org)

### Changelog

For all notable changes see [CHANGELOG](https://github.com/elegoev/basebox-ubuntu-18.04-chartmuseum/blob/master/CHANGELOG.md)

### License

Licensed under The MIT License (MIT) - for the full copyright and license information, please view the [LICENSE](https://github.com/elegoev/basebox-ubuntu-18.04-chartmuseum/blob/master/LICENSE) file.

### Issue Reporting

Any and all feedback is welcome.  Please let me know of any issues you may find in the bug tracker on github. You can find it [here.](https://github.com/elegoev/basebox-ubuntu-18.04-chartmuseum/issues)


CHANGELOG.md
# Change Log

## 1.0.x

**Implemented enhancements:**

- chartmuseum installation

## 2.0.x

**Implemented enhancements:**

- Packer build implemented


boxtests/docker_spec.rb
# Test Docker
describe docker_container('chartmuseum') do
  it { should be_running }
  # matcher syntax: have_volume(container_path,host_path)
  it { should have_volume('/charts','/home/vagrant/charts') }
end


boxtests/vagrant_spec.rb
# Test Vagrant Image
describe 'Vagrant Image Tests' do
  describe user('vagrant') do
    it { should exist }
    it { should belong_to_group 'vagrant' }
  end
end


packer/provisioners/base-provisioning.sh
#!/bin/bash

application_file_path="/vagrant/installed-application.md"

# install chartmuseum
IMAGETAG="0.12.0"
LOCALDIR="/home/vagrant/charts"
echo "install chartmuseum version $IMAGETAG"
sudo docker run -d --restart=unless-stopped --name chartmuseum -p 8080:8080 \
     -v $LOCALDIR:/charts \
     -e DEBUG=true \
     -e STORAGE=local \
     -e STORAGE_LOCAL_ROOTDIR=/charts \
     chartmuseum/chartmuseum:v$IMAGETAG

# change owner & group
sudo chown -R vagrant $LOCALDIR
sudo chgrp -R vagrant $LOCALDIR

# set version
DOCKER_VERSION=$(sudo docker version --format '{{.Server.Version}}')
CHARTMUSEUM_VERSION=$IMAGETAG
echo "# Installed application "  > $application_file_path
echo "***                     " >> $application_file_path
echo "> Docker:      $DOCKER_VERSION" >> $application_file_path
echo "> Chartmuseum: $CHARTMUSEUM_VERSION" >> $application_file_path




packer/post-processors/create-git-tag.ps1
# create git tag

$src_image_version = $env:IMAGE_VERSION

$vagrant_file_template = '
Vagrant.configure("2") do |config|

    $basebox_name="ubuntu-18.04-chartmuseum-test"
    $basebox_hostname="ubuntu-1804-chartmuseum-test"
    $src_image_name="elegoev/ubuntu-18.04-chartmuseum"
    $vb_group_name="basebox-chartmuseum-test"

    config.vm.define "#{$basebox_name}" do |machine|
      machine.vm.box = "#{$src_image_name}"
      machine.vm.box_version = "$image_version"
    
      # define guest hostname
      machine.vm.hostname = "#{$basebox_hostname}"

      machine.vm.provider "virtualbox" do |vb|
        vb.name = $basebox_name
        vb.cpus = 1
        vb.customize ["modifyvm", :id, "--memory", "1024" ]
        vb.customize ["modifyvm", :id, "--natdnshostresolver1", "on"]
        vb.customize ["modifyvm", :id, "--natdnsproxy1", "on"]
        vb.customize ["modifyvm", :id, "--groups", "/#{$vb_group_name}" ]
        vb.customize ["modifyvm", :id, "--vram", 256 ]
      end

    end   

end
'

# create Vagrantfile for download test
$vagrant_file_expanded = $vagrant_file_template.Replace('$image_version', "$src_image_version")
Set-Content -Path './test/download/Vagrantfile' -Value "$vagrant_file_expanded"

# create git tag
git add --all
git commit -m "new basebox version $src_image_version"
git push 
git tag $src_image_version
git push --tags





